implementation/egads/src/main/java/net/opentsdb/query/anomaly/egads/olympicscoring/OlympicScoringNode.java

```diff
@@ -0,0 +1,815 @@
+        if (t instanceof Exception) {
+          handleError((Exception) t, true);
+        } else {
+          handleError(new RuntimeException(t), true);
+        }
+      } else {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Failure in baseline query after initial failure", t);
+        }
+      }
+    }
+    
+  }
+    
+  class SubQueryCB implements Callback<Void, Void> {
+    final QueryContext context;
+    
+    SubQueryCB(final QueryContext context) {
+      this.context = context;
+    }
+    
+    @Override
+    public Void call(final Void arg) throws Exception {
+      context.fetchNext(null);
+      return null;
+    }
+    
+  }
+  
+  class ErrorCB implements Callback<Void, Exception> {
+
+    @Override
+    public Void call(final Exception e) throws Exception {
+      if (failed.compareAndSet(false, true)) {
+        onError(e);
+      } else {
+        LOG.warn("Failure launching baseline query after initial failure", e);
+      }
+      return null;
+    }
+    
+  }
+
+  void fetchBaselineData() {
+    // see we should actually start by checking the state cache.
+    if (!startPrediction()) {
+      return;
+    }
+    
+    baseline_queries = new BaselineQuery[config.getBaselineNumPeriods()];
+    final TimeStamp start = new SecondTimeStamp(prediction_start);
+    final TemporalAmount period = DateTime.parseDuration2(config.getBaselinePeriod());
+    
+    // advance to the oldest time first
+    final TimeStamp end = context.query().endTime().getCopy();
+    for (int i = 0; i < config.getBaselineNumPeriods(); i++) {
+      start.subtract(period);
+      end.subtract(period);
+    }
+    if (config.getMode() != ExecutionMode.CONFIG) {
+      end.update(start);
+      end.add(Duration.of(1, model_units));
+    }
+    
+    // fire!
+    for (int i = 0; i < config.getBaselineNumPeriods(); i++) {
+      final BaselineQuery query = new BaselineQuery(i);
+      baseline_queries[i] = query;
+      query.sub_context = buildQuery((int) start.epoch(), 
+                                     (int) end.epoch(), 
+                                     context.queryContext(), 
+                                     query);
+      start.add(baseline_period);
+      end.add(baseline_period);
+    }
+
+    baseline_queries[0].sub_context.initialize(null)
+      .addCallback(new SubQueryCB(baseline_queries[0].sub_context))
+      .addErrback(new ErrorCB());
+    
+    if (config.getMode() == ExecutionMode.PREDICT) {
+      // return here.
+      final AnomalyPredictionState state = cache.getState(cache_key);
+      QueryExecutionException e = new QueryExecutionException("Successfully "
+          + "started prediction start [" + prediction_start + "] and key " 
+          + Arrays.toString(cache_key) + " State: " + JSON.serializeToString(state), 423);
+      sendUpstream(e);
+    }
+  }
+  
+  void countdown() {
+    latch.countDown();
+    if (latch.getCount() == 0) {
```

Comment:
The `getCount()` method is not `synchronized` or otherwise synchronized. If you want to use this for control, which it seems you do, then we need a different method.

See: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CountDownLatch.html#getCount--

Reply:
Yeah I can switch to an atomic int instead, I always start with these just so I know what's going on when roughing it out.